### PR TITLE
Xen orchestra inventory plugin

### PR DESCRIPTION
**Pull request by shinuza**
##### SUMMARY

This adds an ansible inventory plugins that lists and group by types the following assets:

- Virtual machines
- Hosts
- Pools

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
xen_orchestra

##### ADDITIONAL INFORMATION
As vm's are not guaranteed to have an ip [1], I opted to use the assets uuid's instead, as they are consistent across all asset types.
This is lacking tests, although I have a fair understanding of what kind of tests need to be added to this PR, I don't get how to run them. 😄 
Also: I will cleanup the commit history as needed.

[1] A vm without the xen tools or a vm in a `Shutdown` state will not have an ip.

### CODE REVIEW

File: plugins/inventory/xen_orchestra.py [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+import asyncio
+import json
+import os
+import re
+
+
+DOCUMENTATION = '''
+    name: xen_orchestra
+    short_description: Xen Orchestra inventory source
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:17]:
```suggestion
    short_description: Xen Orchestra inventory source
    version_added: 3.7.0
```

File: plugins/inventory/xen_orchestra.py [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+    author:
+        - Dom Del Nano (@ddelnano) <ddelnano@gmail.com>
+        - Samori Gorse (@shinuza) <samorigorse@gmail.com>
+    requirements:
+        - jsonrpc_websocket >= 3.0
+    description:
+        - Get inventory hosts from a Xen Orchestra deployment.
+        - 'Uses a configuration file as an inventory source, it must end in C(.xen_orchestra.yml) or C(.xen_orchestra.yaml)'
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:25]:
```suggestion
        - 'Uses a configuration file as an inventory source, it must end in C(.xen_orchestra.yml) or C(.xen_orchestra.yaml).'
```
All descriptions should always end with a period (or another punctuation mark).

File: plugins/inventory/xen_orchestra.py [hunk 3]
```diff
@@ -0,0 +1,259 @@
...
+    extends_documentation_fragment:
+        - constructed
+        - inventory_cache
+    options:
+        plugin:
+            description: The name of this plugin, it should always be set to C(community.general.xen_orchestra) for this plugin to recognize it as it's own.
+            required: yes
+            choices: ['xen_orchestra', 'community.general.xen_orchestra']
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:33]:
```suggestion
            choices: ['community.general.xen_orchestra']
```
The short form should only be allowed for plugins which were present in Ansible 2.9.

File: plugins/inventory/xen_orchestra.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+            type: str
+        api_host:
+            description:
+                - Api host to XOA api.
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:37]:
```suggestion
                - API host to XOA API.
```
API (and other abbreviations) should always be capitalized.

File: plugins/inventory/xen_orchestra.py [hunk 5]
```diff
@@ -0,0 +1,259 @@
...
+                - If the value is not specified in the inventory configuration, the value of environment variable C(XO_API_HOST) will be used instead.
+            default: '192.168.1.123'
+            type: str
+            env:
+                - name: XO_API_HOST
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:42]: Are these official Xen environment variable names? If they are specific to this inventory plugin (and not to Xen as well), they should be prefixed with `ANSIBLE_`.

File: plugins/inventory/xen_orchestra.py [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+            env:
+                - name: XO_PASSWORD
+        validate_certs:
+            description: Verify SSL certificate if using HTTPS.
+            type: boolean
+            default: yes
+        use_ssl:
+            description: Use wss when connecting to the Xen Orchestra API
+            type: boolean
+            default: yes
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:66]:
```suggestion
            default: true
```
Always use `true` and `false` as YAML booleans (appears more than once).

File: plugins/inventory/xen_orchestra.py [hunk 7]
```diff
@@ -0,0 +1,259 @@
...
+SUSPENDED = 'Suspended'
+POWER_STATES = [RUNNING, HALTED, SUSPENDED, PAUSED]
+HOST_GROUP = 'xo_hosts'
+POOL_GROUP = 'xo_pools'
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+    ''' Host inventory parser for ansible using XenOrchestra as source. '''
+
+    NAME = 'xoa'
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:94]:
```suggestion
    NAME = 'community.general.xen_orchestra'
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added a first few comments. Please note that you need to add a BOTMETA entry and some form of tests (unit or integration tests), as that's required for all new plugins and modules (see also https://github.com/ansible-collections/community.general/blob/main/CONTRIBUTING.md#creating-new-modules-or-plugins).

**Comment by Ajpantuso** [comment]
Here are some links to the docs as well to help you resolve the `sanity-check` failures.

[forward-compatibility boilerplate](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#use-forward-compatibility-boilerplate)
[str.format()](https://docs.ansible.com/ansible/latest/dev_guide/developing_python_3.html#string-formatting-for-controller-code)
[imports](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#python-imports)
[third-party imports](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_best_practices.html#importing-and-using-shared-code)

I didn't find reference to it, but typically imports are ordered as follows:
1. Standard Library modules
2. Internal Ansible modules
3. third-party packages (Enclosed in `try/except` blocks)

**Comment by felixfontein** [comment]
This one is also helpful for third-party imports: https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/dev_guide/testing/sanity/import.rst

File: plugins/inventory/xen_orchestra.py (hunk 5 above)
Comment (shinuza) [reply plugins/inventory/xen_orchestra.py]: This is not something standard in the Xen env, I prefixed them

**Comment by felixfontein** [comment]
@shinuza are you still working on this?

needs_info

File: plugins/inventory/xen_orchestra.py [hunk 8]
```diff
@@ -0,0 +1,303 @@
...
+# Copyright (c) 2021 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: xen_orchestra
+    short_description: Xen Orchestra inventory source
+    version_added: 4.0.0
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:11]:
```suggestion
    version_added: 4.1.0
```

File: plugins/inventory/xen_orchestra.py [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+        - websocket-client >= 1.0.0
+    description:
+        - Get inventory hosts from a Xen Orchestra deployment.
+        - 'Uses a configuration file as an inventory source, it must end in C(.xen_orchestra.yml) or C(.xen_orchestra.yaml).'
+    extends_documentation_fragment:
+        - constructed
+        - inventory_cache
+    options:
+        plugin:
+            description: The name of this plugin, it should always be set to C(community.general.xen_orchestra) for this plugin to recognize it as it's own.
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:25]:
```suggestion
            description: The name of this plugin, it should always be set to C(community.general.xen_orchestra) for this plugin to recognize it as its own.
```

File: plugins/inventory/xen_orchestra.py [hunk 10]
```diff
@@ -0,0 +1,303 @@
...
+        password:
+            description:
+                - Xen Orchestra password.
+                - If the value is not specified in the inventory configuration, the value of environment variable C(ANSIBLE_XO_PASSWORD) will be used instead.
+            required: yes
+            type: str
+            env:
+                - name: ANSIBLE_XO_PASSWORD
+        validate_certs:
+            description: Verify SSL certificate if using HTTPS.
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:53]:
```suggestion
            description: Verify TLS certificate if using HTTPS.
```
Or write `TLS/SSL` if you think people find the mention of SSL helpful.

File: plugins/inventory/xen_orchestra.py [hunk 11]
```diff
@@ -0,0 +1,303 @@
...
+SUSPENDED = 'Suspended'
+POWER_STATES = [RUNNING, HALTED, SUSPENDED, PAUSED]
+HOST_GROUP = 'xo_hosts'
+POOL_GROUP = 'xo_pools'
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+    ''' Host inventory parser for ansible using XenOrchestra as source. '''
+
+    NAME = 'xen_orchestra'
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:107]:
```suggestion
    NAME = 'community.general.xen_orchestra'
```

File: plugins/inventory/xen_orchestra.py [hunk 12]
```diff
@@ -0,0 +1,303 @@
...
+            if path.endswith(('xen_orchestra.yaml', 'xen_orchestra.yml')):
+                valid = True
+            else:
+                self.display.vvv(
+                    'Skipping due to inventory source not ending in "xen_orchestra.yaml" nor "xen_orchestra.yml"')
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        if not HAS_WEBSOCKET:
+            raise AnsibleError('This module requires websocket-client 1.0.0 or higher: '
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:282]:
```suggestion
            raise AnsibleError('This plugin requires websocket-client 1.0.0 or higher: '
```

File: plugins/inventory/xen_orchestra.py [hunk 13]
```diff
@@ -0,0 +1,303 @@
...
+        self._add_hosts(objects['hosts'], objects['pools'])
+        self._add_vms(objects['vms'], objects['hosts'], objects['pools'])
+
+    def to_safe(self, word):
+        '''Converts 'bad' characters in a string to underscores so they can be used as Ansible groups
+        #> InventoryModule.to_safe("foo-bar baz")
+        'foo_barbaz'
+        '''
+        regex = r"[^A-Za-z0-9\_]"
+        return re.sub(regex, "_", word.replace(" ", "")).lower()
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:267]: Ansible already has a built-in (and configurable) mechanism to sanitize group names. Have you considered not doing this in this plugin, but simply let Ansible deal with it?

Comment (shinuza) [reply plugins/inventory/xen_orchestra.py:267]: I didn't know. I'm touching the tests up and I'll test this as well. Thanks for the review/heads up

Comment (shinuza) [reply plugins/inventory/xen_orchestra.py]: I'm not relying on built-in sanitizer, although I add to replace `_` and ` ` (spaces) before running them through

File: plugins/inventory/xen_orchestra.py [hunk 14]
```diff
@@ -22,9 +22,9 @@
         - inventory_cache
     options:
         plugin:
-            description: The name of this plugin, it should always be set to C(community.general.xen_orchestra) for this plugin to recognize it as its own.
+            description: The name of this plugin, it should always be set to C(xen_orchestra) for this plugin to recognize it as its own.
             required: yes
-            choices: ['community.general.xen_orchestra']
+            choices: ['xen_orchestra']
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:27]: Can you change this back to `community.general.xen_orchestra`? Thanks.

Comment (shinuza) [reply plugins/inventory/xen_orchestra.py]: Sorry, I was testing a bit of code and I pushed this by accident

File: plugins/inventory/xen_orchestra.py [hunk 15]
```diff
@@ -0,0 +1,315 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2021 Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = '''
+    name: community.general.xen_orchestra
```
Comment (felixfontein) [inline plugins/inventory/xen_orchestra.py:9]:
```suggestion
    name: xen_orchestra
```
This one needs to stay short. (It's the only one :) )

**Review by felixfontein — APPROVED** [review]
Looks good to me.

**Comment by felixfontein** [comment]
Will merge if nobody complains in, say, a week.

**Comment by felixfontein** [comment]
@shinuza thanks for contributing this plugin!
@Ajpantuso thanks for reviewing!

**Comment by shinuza** [comment]
Thank you, more to come!

**Pull request merged**